kgforge/core/archetypes/external_dataset.py

```diff
@@ -0,0 +1,134 @@
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Blue Brain Nexus Forge. If not, see <https://choosealicense.com/licenses/lgpl-3.0/>.
+
+import json
+import requests
+from abc import abstractmethod
+
+from typing import Optional, Union, List
+from kgforge.core import Resource
+from kgforge.core.archetypes.read_store import ReadStore
+from kgforge.core.archetypes.model import Model
+from kgforge.core.commons.context import Context
+from kgforge.core.commons.exceptions import QueryingError
+from kgforge.specializations.stores.bluebrain_nexus import BlueBrainNexus
+
+
+class ExternalDataset(ReadStore):
+    """A class to link to external databases, query and search directly on datasets. """
+
+    def __init__(self, model: Optional[Model] = None,
+                 ) -> None:
+        super().__init__(model)
+
+    def types(self):
+        # TODO: add other datatypes used, for instance, inside the mappings
+        return list(self.model.mappings(self.model.source, False).keys())
+
+    def search(self, resolvers, *filters, **params):
+        """Search within the database.
+
+        :param keep_original: bool
+        """
+        keep_original = params.pop('keep_original', True)
+        unmapped_resources = self._search(resolvers, *filters, **params)
+        if isinstance(self.service, BlueBrainNexus) or keep_original:
+            return unmapped_resources
+        # Try to find the type of the resources within the filters
+        resource_type = type_from_filters(*filters)
+        return self.map(unmapped_resources, type_=resource_type)
+
+    @abstractmethod
+    def _search(self):
+        ...
+
+    def sparql(self, query: str, debug: bool = False, limit: Optional[int] = None,
+               offset: Optional[int] = None, **params) -> Optional[Union[List[Resource], Resource]]:
+        """Use SPARQL within the database.
+
+        :param keep_original: bool
+        """
+        keep_original = params.pop('keep_original', True)
+        unmapped_resources = self._sparql(query, debug, limit, offset, **params)
+        if keep_original:
+            return unmapped_resources
+        return self.map(unmapped_resources)
+
+
+def type_from_filters(*filters) -> Optional[str]:
+    """Returns the first `type` found in filters."""
+    resource_type = None
+    filters = filters[0]
+    if isinstance(filters, dict):
+        if 'type' in filters:
+            resource_type = filters['type']
+    else:
+        # check filters grouping
+        if isinstance(filters, (list, tuple)):
+            filters = [filter for filter in filters]
+        else:
+            filters = [filters]
+        for filter in filters:
+            if 'type' in filter.path and filter.operator is "__eq__":
+                resource_type = filter.value
+                break
+    return resource_type
+
+
+def resources_from_request(url, headers, **params):
```

Comment:
I did not see being used. I recommend to add it later once it is used. Meanwhile, I think this can go to utils. It seems yet not quite generic ('response_loc'). One way to handle this could be to make the request and then call a callback method for result handling. The caller provide the callback method